beginner_source/aws_distributed_training_tutorial.py

```diff
@@ -0,0 +1,675 @@
+val_loader = torch.utils.data.DataLoader(valset, batch_size=batch_size, shuffle=False, num_workers=workers, pin_memory=False)
+
+
+######################################################################
+# Training Loop
+# ~~~~~~~~~~~~~
+# 
+# The last step is to define the training loop. We have already done most
+# of the work for setting up the distributed training so this is not
+# distributed training specific. The only detail is setting the current
+# epoch count in the ``DistributedSampler``, as the sampler shuffles the
+# data going to each process deterministically based on epoch. After
+# updating the sampler, the loop runs a full training epoch, runs a full
+# validation step then prints the performance of the current model against
+# the best performing model so far. After training for num\_epochs, the
+# loop exits and the tutorial is complete. Notice, since this is an
+# exercise we are not saving models but one may wish to keep track of the
+# best performing model then save it at the end of training (see
+# `here <https://github.com/pytorch/examples/blob/master/imagenet/main.py#L184>`__).
+# 
+
+best_prec1 = 0
+
+for epoch in range(num_epochs):
+    # Set epoch count for DistributedSampler
+    train_sampler.set_epoch(epoch)
+    
+    # Adjust learning rate according to schedule
+    adjust_learning_rate(starting_lr, optimizer, epoch)
+
+    # train for one epoch
+    print("\nBegin Training Epoch {}".format(epoch+1))
+    train(train_loader, model, criterion, optimizer, epoch)
+
+    # evaluate on validation set
+    print("Begin Validation @ Epoch {}".format(epoch+1))
+    prec1 = validate(val_loader, model, criterion)
+
+    # remember best prec@1 and save checkpoint if desired
+    #is_best = prec1 > best_prec1
+    best_prec1 = max(prec1, best_prec1)
+
+    print("Epoch Summary: ")
+    print("\tEpoch Accuracy: {}".format(prec1))
+    print("\tBest Accuracy: {}".format(best_prec1))
+
+
+######################################################################
+# Running the Code
+# ----------------
+# 
+# Unlike most of the other PyTorch tutorials, this code may not be run
+# directly out of this notebook. To run, download the .py version of this
+# file (or convert it using
+# `this <https://gist.github.com/chsasank/7218ca16f8d022e02a9c0deb94a310fe>`__)
+# and upload a copy to both nodes. The astute reader would have noticed
+# that we hardcoded the **node0-privateIP** and :math:`world\_size=4` but
+# input the *rank* and *local\_rank* inputs as arg[1] and arg[2] command
+# line arguments, respectively. Once uploaded, open two ssh terminals into
+# each node.
+# 
+# -  On the first terminal for node0, run ``$ python main.py 0 0``
+# 
+# -  On the second terminal for node0 run ``$ python main.py 1 1``
+# 
+# -  On the first terminal for node1, run ``$ python main.py 2 0``
+# 
+# -  On the second terminal for node1 run ``$ python main.py 3 1``
```

Comment:
I would prefer you add an argument parser for the main.py and avoid using these 0 0  1 1 arguments. They should come with a name to be more clear to the user which number means what.